File: C++/ArrayString/27_3SumClosest.cpp
#include <bits/stdc++.h>
using namespace std;

int threeSumClosest(vector<int> &nums, int target)
{
    if (nums.size() < 3)
        return 0;
    sort(nums.begin(), nums.end());
    int diff = INT_MAX;
    int ans = INT_MIN;
    for (int i = 0; i < nums.size() - 2; i++)
    {
        if (i == 0 or (i > 0 and nums[i] != nums[i - 1]))
        {
            int si = i + 1, ei = nums.size() - 1;
            while (si < ei)
            {
                int currDiff = abs(target - (nums[i] + nums[si] + nums[ei]));
                if (currDiff < diff)
                {
                    diff = currDiff;
                    ans = nums[i] + nums[si] + nums[ei];
                }
                else if (nums[i] + nums[si] + nums[ei] < target)
                    si++;
                else
                    ei--;
            }
        }
    }
    return ans;
}

File: Java/ArrayString/27_3SumClosest.java
public int twoSumClosest(int[] arr,int si,int ei,int tar){
        int ans = (int)1e9;
        while(si < ei){
            int sum = arr[si] + arr[ei];
            if(sum == tar){
                return tar;
            }
            else if(sum < tar){
                if(tar - sum < ((ans > tar)?ans-tar:tar-ans) ) ans = sum;
                si++;
            }
            else{
                if(sum - tar < ((ans > tar)?ans-tar:tar-ans)  ) ans = sum;
                ei--;
            }
        }
        return ans;
    }
    public int threeSumClosest(int[] arr, int tar) {
        Arrays.sort(arr);
        int ans = (int)1e9;
        int n =arr.length;
        for(int i=0;i<n-2;i++){
            int smallAns = twoSumClosest(arr,i+1,n-1,tar - arr[i]);            
            smallAns = smallAns + arr[i];
            if(smallAns == tar) return tar;
            else if(smallAns < tar){
                ans = (tar - smallAns < ((ans > tar)?ans-tar:tar-ans)  ) ? smallAns : ans;
            }
            else{
                ans = (smallAns - tar < ((ans > tar)?ans-tar:tar-ans)  ) ? smallAns : ans;
            }
        }
        return ans;
    }